### DISCUSSION TITLE
Adapt NFCGate to operate within a virtual environment

### DISCUSSION THREAD

**Discussion by mohaomer12** [nfcgate/nfcgate#135] [Ideas]
Hello,

Is it theoretically possible to adapt NFCGate to operate within a virtual environment like OpenSDK, rather than relying on a rooted phone? Utilizing a rooted phone presents numerous challenges, including significant effort and constraints on compatible phone hardware for NFCGate.

https://github.com/WaxMoon/opensdk/tree/eaab1645e57c0be6d3c1b0f5bf586984f4de141f

**Comment by kleest** [comment]
I don't know OpenSDK. Could you point me to some documentation please? Unfortunately, the repository does not contain documentation.

**Reply by mohaomer12** [reply]
Here an example project using OpenSDK to run multiples apps in the virtual environment with OpenSDK API examples:
https://github.com/WaxMoon/MultiApp
And here an example for a virtual camera project  that replace the hardware camera data acquisition with prerecorded video without rooting the phone:
https://github.com/andvipgroup/VCamera

**Reply by kleest** [reply]
It looks like the VCamera projects is incomplete, since it does not contain any camera replacement logic but only the management code (resizing, select source, etc.). However, as far as I understand it, target apps are placed in a virtual environment and this virtual environment then replaces runtime APIs.

Going back to NFCGate, I don't think this would benefit NFCGate much since the idea of this project is to use a smartphone as an NFC toolkit with existing tags and readers and not target apps on the running device directly. The only operation mode this could benefit is the on-device-capture mode, which is different from the other modes: It hooks into the NFC runtime APIs to capture NFC traffic on the devices itself. Of course it may also be used to feed modified NFC traffic to apps on the device, however this is not yet implemented. In this mode, some kind of virtualization could be useful to remove the Xposed requirement. But, I personally prefer to have some "Xposed compatibility layer" for opensdk so that NFCGate can just use proven and stable Xposed instead of implementing opensdk logic for launching apps.
